blog/_src/posts/2017-11-XX-adding-refinement-types.md

```diff
@@ -0,0 +1,449 @@
+```
+
+It states that the index (the second argument) must be an `Integer`,
+i.e. any whole-number value in the exclusive range (-∞ , ∞). This
+description, however, is somewhat coarse: we obviously cannot use
+_any_ integer. To avoid a runtime error when indexing into a vector
+`v`, we must use an integer in the range [0, `(vector-length
+v)`). With a refinement type we can precisely describe this subset of
+the integers:
+
+```racket
+(Refine [i : Integer] (<= 0 i (- (vector-length v) 1)))
+```
+
+This refinement type describes values `i` of type `Integer` such that
+`(<= 0 i (- (vector-length v) 1))` produces `#true`. The documentation
+for [Refine](http://docs.racket-lang.org/search/index.html?q=Refine)
+describes precisely what propositions can be used in
+refinements. Today it includes logical statements about the types of
+terms, standard logical combinators (e.g. and, or), and linear integer
+constraints.
+
+## Dependent function types
+
+By combining refinement types with dependent function types
+(i.e. functions whose domain and range may depend on argument and/or
+program values), we can now give a more precise function type for many
+basic vector operations, e.g.:
+
+```racket
+(: safe-vector-ref 
+   (All (A) (-> ([v : (Vectorof A)]
+                 [n : (v) (Refine [i : Integer]
+                                  (<= 0 i (- (vector-length v) 1)))])
+                A)))
+(define safe-vector-ref vector-ref)
+```
+
+Here we have stated that `safe-vector-ref`'s second argument (`n`)
+depends on its first argument (`v`) and that the type of `n` is an
+`Integer` which is a valid index for `v`.
+
+If we try and use `safe-vector-ref` with a vector and index which
+Typed Racket cannot prove satisfy the respective types:
+
+```racket
+(safe-vector-ref (vector 0 1) 2)
+```
+
+Typed Racket will report an error while type checking:
+
+```
+Type Checker: type mismatch
+  expected: (Refine
+             (i : Integer)
+             (and (<= i (- (vector-length v) 1)) (<= 0 v)))
+  given: (Refine (z : Positive-Byte) (= 2 z)) in: 2
+```
+
+### Preconditions
+
+Sometimes the refinements used in a dependent function type can also
+be written conveniently as a "precondition", e.g.:
+
+```racket
+(: safe-vector-ref
+   (All (A) (-> ([v : (Vectorof A)]
+                 [n : Integer])
+                #:pre (v n) (<= 0 n (- (vector-length v) 1))
+                A)))
+(define safe-vector-ref vector-ref)
+```
+
+This function type is equivalent to the previous version, but the
+refining proposition that requires that `n` be a valid index for `v`
+now directly references `n` and has been written directly after the
+list of arguments:
+
+```racket
+#:pre (v n) (<= 0 n (- (vector-length v) 1))
+```
+
+We can read the precondition specification above as follows: the
+precondition for the function depends on arguments `v` and `n` and it
+requires that Typed Racket can prove `(<= 0 n (- (vector-length v)
+1))` would produce `#true` at all call sites.
```

Comment:
This sounds backwards to me, like "you can only write this type if TR can prove something about all call sites"

suggestion:
> the precondition for `safe-vector-ref` depends on arguments `v` and `n` and states that a call `(safe-vector-ref v n)` is well typed only if `(<= 0 n (- (vector-length v))`